### ISSUE TITLE
How do deal with distributions?

### ISSUE THREAD

**Issue by probonopd** [AppImage/libappimage#4]
How do deal with distributions?

By the time libappimage lands in debian stable or the oldest still-supported Ubuntu LTS, we will long have moved on.

Can we somehow enforce that every user of this library bundles their own private copy, ensuring that once this is in debian stable or the oldest still-supported Ubuntu LTS, people don't use that outdated version but the most recent one?

**Comment by Conan-Kudo** [comment]
The smarter thing to do would be to validate the ABI. If the ABI remains validated, even Debian will be willing to ship updates of the library as regular updates (as opposed to backports).

**Comment by Conan-Kudo** [comment]
Also, if this is a "core library" implementing AppImage, it _should not_ be changing anywhere near as much as the rest of your stuff. It's supposed to be predictable and usable.

Both Flatpak and Snappy use similar strategies with their base libraries (libflatpak and snapd-glib), which allows for distributions to keep updating it because they have a high degree of confidence that it won't break anything.

**Comment by shoogle** [comment]
Here's how Qt maintains ABI compatibility: https://wiki.qt.io/D-Pointer

That's probably the best approach to take, but if you wanted to really annoy the distributions you could always take a leaf out of Python `pip`'s book and make it self-updating. 😉

**Comment by TheAssassin** [comment]
They basically utilize the opaque data class pattern which I have introduced in most of my (new C++) code in other AppImage related projects. I've been telling about its advantages for quite a while. Moving all internal state stuff to the implementation provides lots of benefits (which outweigh the cons by far, which is mostly you have to dereference a pointer when accessing state variables (and that overhead is really small, compared to what most methods do)).

**Comment by azubieta** [comment]
Opaque pointer pattern is already being applied to the new implementation.

**Comment by probonopd** [comment]
D-Pointer, too?

**Comment by azubieta** [comment]
@probonopd D-Pointer is a variation of Opaque pointer. But yes it's implemented that way.

**Comment by TheAssassin** [comment]
Not quite. D-Pointer is a specialization (as in, names a few things so they're used consistently, e.g., `d_ptr` member variable) of the opaque data class pattern. The opaque pointer pattern is *another* specialization of the opaque data class pattern, mostly used in C. D-Pointer is *not* a specialization of opaque pointer or vice versa.

**Comment by probonopd** [comment]
Is this the kind of stuff one learns in Comp Sci graduate courses?

**Comment by TheAssassin** [comment]
It's called Wikipedia.

**Comment by azubieta** [comment]
I would like to clarify that the greater issue here is not how to keep backward compatibility but how to ship updates to those using the old binaries.

**Comment by probonopd** [comment]
> Sadly it's a thing that will run out of our hands.

The day will come where we will have to discourage distributions from shipping our stuff, because most will ship updated stuff. (Think Debian stable.)
